examples/backends/bls/src/bls.cc

```diff
@@ -0,0 +1,621 @@
+  std::vector<std::string> model_names = {"addsub_python", "addsub_tf"};
+
+  // The way we collect these batch timestamps is not entirely
+  // accurate. Normally, in a performant backend you would execute all
+  // the requests at the same time, and so there would be a single
+  // compute-start / compute-end time-range. But here we execute each
+  // request separately so there is no single range. As a result we
+  // just show the entire execute time as being the compute time as
+  // well.
+  uint64_t compute_start_ns = 0;
+  SET_TIMESTAMP(compute_start_ns);
+
+  // Process requests
+  for (size_t r = 0; r < request_count; r++) {
+    try {
+      for (size_t i = 0; i < 2; i++) {
+        // Check if the model is ready.
+        bool is_ready = false;
+        THROW_IF_TRITON_ERROR(TRITONSERVER_ServerModelIsReady(
+            model_state->TritonServer(), model_names[i].c_str(),
+            1 /* model_version */, &is_ready));
+        if (!is_ready) {
+          throw BLSBackendException(
+              (std::string("Failed to execute the inference request. Model '") +
+               model_names[i].c_str() + "' is not ready.")
+                  .c_str());
+        }
+
+        // Fail to execute the inference if the model is using the decoupled
+        // transaction policy.
+        uint32_t txn_flags;
+        THROW_IF_TRITON_ERROR(TRITONSERVER_ServerModelTransactionProperties(
+            model_state->TritonServer(), model_names[i].c_str(),
+            1 /* model_version */, &txn_flags, nullptr /* voidp */));
+
+        // Decoupled API is not supported in the current BLS interface.
+        if ((txn_flags & TRITONSERVER_TXN_DECOUPLED) != 0) {
+          throw BLSBackendException(
+              std::string("Model '") + model_names[i].c_str() +
+              "' is using the decoupled. BLS doesn't support models using the "
+              "decoupled transaction policy.");
+        }
+      }
+    }
+    catch (const BLSBackendException& bls_exception) {
+      LOG_MESSAGE(TRITONSERVER_LOG_ERROR, bls_exception.what());
+      RESPOND_AND_SET_NULL_IF_ERROR(
+          &responses[r],
+          TRITONSERVER_ErrorNew(
+              TRITONSERVER_ERROR_INTERNAL,
+              "Failed to send inference requests, models are not ready"));
+      break;
+    }
+
+    // For each inference request, the backend sends two requests on the
+    // 'addsub_python' and 'addsub_tf' models.
+    // Prepare std::future for each model. Since this BLS backend
+    // can handle requests in parallel, we will send all the inference
+    // requests first and then retrieve them later.
+    std::vector<std::future<TRITONSERVER_InferenceResponse*>> futures(2);
+
+    // The inference request object for sending internal requests.
+    TRITONSERVER_InferenceRequest* irequest = nullptr;
+
+    // Create a BLSExecutor object. The custom BLS logic is placed inside
+    // class BLSExecutor.
+    BLSExecutor* bls_executor = nullptr;
```

Comment:
There is memory leak because bls_executor is never deleted. You can just use a local variable.